### ISSUE TITLE
opt qwen

### ISSUE THREAD

**Issue by Wang-tiantian** [Wang-tiantian/test#17]
import os
# Setup environment (full fine-tuning 7b needs 160GB Memor
import torch
import pandas as pd
from datasets import load_dataset, Dataset, DatasetDict
from transformers import AutoModelForCausalLM, AutoTokenizer, EarlyStoppingCallback, TrainingArguments
import logging
from peft import LoraConfig, PeftModel
from trl import SFTTrainer
from transformers import TrainerCallback
print(torch.cuda.device_count())
import nltk
import Levenshtein
import evaluate
from nltk.translate.bleu_score import sentence_bleu, SmoothingFunction
from rapidfuzz.distance import Levenshtein

# Data Loading and Preprocessing
train_file = "data/data_for_llms/thermal_train1255.csv" #255 1255
test_file = "data/data_for_llms/thermal_test.csv" #28 248

train_df = pd.read_csv(train_file, encoding='utf-8')
test_df = pd.read_csv(test_file, encoding='utf-8')
print(train_df.info())
print(test_df.info())
# print(test_df['articleInformation'][0])
# print(test_df['reactionDetails'][0])

# Prompt
new_thermal_reaction_details_extraction = """
Your task is to extract the reactionDetails about catalytic reactions from the articleInformation accordingto the mainReactions.
The following is Typescript code that describes the data structure of reactions. The reactions to be exported are not implemented yet.

/*This is a typescript prompt file to export reaction data.*/

// Interface to represent each reactant in the reaction
interface Reactant {
    reactantMixtureRatio?: string; // Mixture ratio (e.g., A:B = x:y)
    reactantList: { // Each reactant in the reaction  
        conversion: string; // Conversion rate of the reactant
        reactantName: string; // Compound name or formula
    }[];
}

// Interface to represent the catalyst used in the reaction
interface Catalyst {
    catalystName: string; // Name or description
    catalystMixtureRatio: string; // Ratio between catalyst and support or catalyst components
    promoter: { // Catalyst promoters
        promoterName: string; // Name or description
        role: string; // Role (e.g., Cocatalyst, Additive)
    }[];
    turnoverFrequency: string; // Turnover frequency, number + unit
    catalystBETSurfaceArea: string; // BET surface area, number + unit
    catalystLoading: string; // Amount of catalyst used
    catalystSupport: string; // Catalyst support material
    catalystShape: string; // Description of catalyst shape and size
    catalystPoreSize: string; // Pore size, number + unit
    catalystPoreVolume: string; // Pore volume, number + unit
    catPreparationMethod: string; // Catalyst preparation method
    catAcidityAndBasicity: string; // Catalyst acidity and basicity
    catMetalOxidationStates: { // Array of metal oxidation states
        metal: string; // Metal name
        oxidationState: number; // Oxidation state
        proportion: string; // Proportion of the metal
    }[];
}

// Interface to represent the reaction conditions
interface Condition {
    reactionType: 'thermal catalysis' | 'photocatalysis' | 'electrocatalysis' | 'enzyme catalysis' | 'others'; // Type of catalytic reaction
    isHeterogeneous: boolean; // Indicates if the reaction is heterogeneous
    temperature: string; // Temperature for achieving optimal yield
    reactionTime: string; // Reaction time for achieving optimal yield
    gases: { 
        pressure: string; // Pressure, number + unit
        gasName: string; // Name of the gas
    }[];
}

// Interface to represent the solvent used in the reaction
interface Solvent {
    solventPH: string; // // Solvent pH (number preferred)
    solventList: { // Each solvent in the reaction
        amount: string; // Solvent amount, number + unit
        solventName: string; // Solvent name
    }[];
}

// Interface to represent each product in the reaction
interface Product {
    yield: string; // Yield of the product, number + unit preferred
    selectivity: string; // Selectivity of the product, number + unit preferred
    productName: string; // Product name or formula
}

// Interface to represent the extracted data for the reaction
interface ExtractedData {
    reactants: Reactant; // Reactants used in the reaction
    catalyst: Catalyst; // Catalyst used in the reaction
    conditions: Condition; // Reaction conditions
    solvents: Solvent; // Solvents used in the reaction
    products: Product[]; // Array of products in the reaction
}

// Interface to represent instructions for the extraction task
interface Instructions {
    generalNote: string; // General note for the task
    notes: string[]; // Array of specific notes for the task
}

const instructions: Instructions = {
    generalNote: "Extract the specified reaction from the provided paper. The given description is a reference, and the actual descriptions of reactants, products, and catalysts in the paper may vary in wording and details. Accurately extract and format relevant information to match the specified reaction. The goal is to focus on the reaction of interest.",
    notes: [
        "Use the paper's information to fill the template below. Extract data strictly according to the original description. If any information is missing or not mentioned, retain its key and assign "unknown".",
        "Each reactant or product should be a separate dictionary in the list, with keys 'reactantName', 'conversion' (for reactants), 'productName', 'yield', and 'selectivity' (for products). If mixture ratios are provided, include them as a single string under 'reactantMixtureRatio' in the format 'A:B = x:y' enclosed in single quotes.",
        "'Conversion' is a property of each reactant, while 'yield' and 'selectivity' are properties of each product. Assign these accurately.",
        "H₂ and O₂ are often key reactants. Ensure these are accurately included.",
           "Default to water (H₂O) as the solvent when "solution" or "aqueous solution" is mentioned unless another solvent is specified."
        "For total values: when only a total value (N%) is given for two products (A and B), apply the total to both products. Add an asterisk * after the value to indicate a shared total (eg. selectivity: "8%*").",
        "Preferentially enter 'yield'. If 'yield' is missing but other values indicating quantity or production rate are available, enter those as 'Number + unit'."
    ]
};

/*This is a reaction data sample to export reaction data.*/

export const reactions: ExtractedData[] = [
    {
        "reactants": {
            "reactantMixtureRatio": "CH4:O2 = 1:1",
            "reactantList" : [
                { "conversion": "90%", "reactantName": "methane" },
                { "conversion": "unknown", "reactantName": "oxygen" }
            ]
        },
        "catalyst": {
            "catalystName": "palladium on carbon",
            "catalystMixtureRatio": "5% Pd on carbon",
            "promoters": [
                { "promoterName": "Iron(III) nitrate", "role": "Cocatalyst" },
                { "promoterName": "Acetic acid", "role": "Additive" }
            ],
            "turnoverFrequency": "1500 h⁻¹",
            "catalystBETSurfaceArea": "1200 m²/g",
            "catalystLoading": "10 g/L",
            "catalystSupport": "carbon black",
            "catalystShape": "Spherical particles, 2.9-3.5 nm",
            "catalystPoreSize": "10 nm",
            "catalystPoreVolume": "0.5 cm³/g",
            "catPreparationMethod": "incipient wetness impregnation",
            "catAcidityAndBasicity": "Highest acidic site concentration: 0.358 mmol/g, total alkalinity: 0.936 mmol/g",
            "catMetalOxidationStates": [
                { "metal": "Platinum", "oxidationState": "0", "proportion:" "95%" },
                { "metal": "Platinum", "oxidationState": 2, "proportion": "unknown" }
            ]
        },
        "conditions": {
            "reactionType": "thermal catalysis",
            "isHeterogeneous": true,
            "temperature": "300°C",
            "reactionTime": "2 hours",
            "gases": [
                { "pressure": "2 bar","gasName": "O2" },
                { "pressure": "1 atm","gasName": "air" }
            ]
        },
        "solvents": {
            "solventPH": "8",
            "solventList" : [
                { "amount": "100 mL", "solventName": "H2O" },
                { "amount": "1.0 mM", "solventName": "Fe(NO3)3" }
            ]
        },
        "products": [
            { "yield": "81%", "selectivity": "90%", "productName": "methanol" },
            { "yield": "unknown", "selectivity": "8%","productName": "Product_A" }
        ]
    }
];

// TODO: export reactions
export const reactions: ExtractedData[] = [
]
"""

train_df['text'] = f'[INST] {new_thermal_reaction_details_extraction} ' + '/* There are articleInformation. */ ' + train_df["articleInformation"] + " [/INST] " + '/*There are mainReactions.*/' + train_df["mainReactions"] + " [/INST] " + train_df["reactionDetails"] + "!!! "
test_df['text'] = f'[INST] {new_thermal_reaction_details_extraction} ' + '/* There are articleInformation. */  ' + test_df["articleInformation"] + " [/INST] " + '/*There are mainReactions.*/' + test_df["mainReactions"] + " [/INST] " + test_df["reactionDetails"] + "!!! "
print(test_df['text'][0])

train_dataset = Dataset.from_dict(train_df[['text']].astype(str))
test_dataset = Dataset.from_dict(test_df[['text']].astype(str))
dataset = DatasetDict({"train": train_dataset, "test": test_dataset})
print(dataset)

# TrainingArguments parameters
num_train_epochs = 2
save_steps = 0 # Save checkpoint every X updates steps
logging_steps = 25 # Log every X updates steps

fp16 = False # Enable fp16/bf16 training (set fp16 to True with an V100)
bf16 = True # Enable fp16/bf16 training (set bf16 to True with an A100)

per_device_train_batch_size = 2 # Batch size per GPU for training
per_device_eval_batch_size = 2 # Batch size per GPU for evaluation
gradient_accumulation_steps = 1 # Number of update steps to accumulate the gradients for
gradient_checkpointing = True # Enable gradient checkpointing

max_grad_norm = 0.3 # Maximum gradient normal (gradient clipping)
learning_rate = 5e-6 # Initial learning rate (AdamW optimizer, 1e-5 or 5e-6 or 1e-4)
weight_decay = 0.001 # Weight decay to apply to all layers except bias/LayerNorm weights

optim = "paged_adamw_32bit" # Optimizer to use
lr_scheduler_type = "cosine" # Learning rate schedule

max_steps = -1 # Number of training steps (overrides num_train_epochs)
warmup_ratio = 0.03 # Ratio of steps for a linear warmup (from 0 to learning rate)

group_by_length = True # Group sequences into batches with same length (Saves memory and speeds up training considerably)

# SFT parameters
max_seq_length = 4096               # Maximum sequence length to use (default 1024)
packing = False                     # Pack multiple short examples in the same input sequence to increase efficiency
device_map = "auto"                 # Load the entire model on the GPU 0, or "auto"

# Model Version (Meta-Llama-3-8B-Instruct, Mistral-7B-Instruct-v0.2, llama-2-13b-chat-hf ...)
model_name = "/public/home/tt/Qwen/Qwen3-8B"  # Path of the pretrained model downloaded from Hugging Face
new_model_dir = f"saved_models/thecatopt2-Qwen3-8B/epoch_{num_train_epochs}_train{len(train_df)}_lr{learning_rate}_bs{per_device_train_batch_size}" # Fine-tuned model name
output_dir = new_model_dir          # Output directory where the model predictions and checkpoints will be stored

# Load base model
model = AutoModelForCausalLM.from_pretrained(
    pretrained_model_name_or_path = model_name,
    torch_dtype = torch.bfloat16,
    device_map = device_map
)

# Load tokenizer
tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)
tokenizer.pad_token = tokenizer.eos_token
tokenizer.padding_side = "right"  # Fix weird overflow issue with fp16 training
print("------------tokenizer.eos_token--------------", tokenizer.eos_token)
print("------------tokenizer.unk_token--------------", tokenizer.unk_token)
print("------------tokenizer.bos_token--------------", tokenizer.bos_token)
print("------------tokenizer.pad_token--------------", tokenizer.pad_token)
print("------------vocab_size is------------", tokenizer.vocab_size)
print("------------vocab_size is------------", len(tokenizer))

# eval
import jieba
import numpy as np
import Levenshtein
from typing import List, Union

def ngram_counts(tokens: List[str], n: int) -> dict:
    """统计n-gram（用元组）出现次数"""
    from collections import Counter
    ngrams = [tuple(tokens[i:i+n]) for i in range(len(tokens)-n+1)]
    return Counter(ngrams)

def modified_precision(candidate: List[str], reference: List[str], n: int) -> float:
    """计算n-gram层次的modified precision"""
    cand_ng = ngram_counts(candidate, n)
    ref_ng = ngram_counts(reference, n)
    overlap = sum(min(count, ref_ng.get(ng, 0)) for ng, count in cand_ng.items())
    total = max(len(candidate) - n + 1, 1)
    return overlap / total if total > 0 else 0.0

def brevity_penalty(c: int, r: int) -> float:
    """BP因子"""
    if c == 0 or r == 0:
        return 0.0
    return 1.0 if c > r else np.exp(1 - r / c)

def compute_bleu(reference: str, candidate: str, max_n: int = 4) -> float:
    """
    自定义BLEU评分计算
    返回: 0.0-1.0之间的分数，无效输入返回0.0
    """
    if not reference or not candidate:
        return 0.0
    
    try:
        ref_tokens = jieba.lcut(reference)
        cand_tokens = jieba.lcut(candidate)
        
        # 处理空分词结果
        if not ref_tokens or not cand_tokens:
            return 0.0
            
        weights = [1.0 / max_n] * max_n
        p_ns = [modified_precision(cand_tokens, ref_tokens, n) for n in range(1, max_n+1)]
        
        # 计算对数加权和
        s = sum(w * np.log(p) if p > 0 else -9999 for w, p in zip(weights, p_ns))
        bp = brevity_penalty(len(cand_tokens), len(ref_tokens))
        
        bleu = bp * np.exp(s)
        return min(max(bleu, 0.0), 1.0)  # 确保在0-1范围内
    except:
        return 0.0

def compute_metrics(eval_pred) -> dict:
    preds, labels = eval_pred.predictions, eval_pred.label_ids
    
    # 处理不同形状的输入
    preds = np.array(preds)
    labels = np.array(labels)
    
    if len(preds.shape) == 3:
        preds = np.argmax(preds, axis=-1)
    if len(labels.shape) == 3:
        labels = np.argmax(labels, axis=-1)
    
    decoded_preds = tokenizer.batch_decode(preds.tolist(), skip_special_tokens=True)
    decoded_labels = tokenizer.batch_decode(labels.tolist(), skip_special_tokens=True)

    bleu_scores = []
    lev_scores = []

    for pred, label in zip(decoded_preds, decoded_labels):
        # 处理None或空字符串
        pred = pred or ""
        label = label or ""
        
        bleu = compute_bleu(label, pred)
        bleu_scores.append(bleu)
        
        # Levenshtein相似度
        lev = Levenshtein.ratio(pred.lower().strip(), label.lower().strip())
        lev_scores.append(lev)

    return {
        "bleu": float(np.mean(bleu_scores)) if bleu_scores else 0.0,
        "levenshtein": float(np.mean(lev_scores)) if lev_scores else 0.0
    }

# Set training parameters
training_arguments = TrainingArguments(
    output_dir = output_dir,
    logging_dir = output_dir + "/logs/",
    # evaluation_strategy = "epoch",
    eval_strategy = "epoch",
    metric_for_best_model="bleu",
    greater_is_better=True,
    load_best_model_at_end=True,
    # logging_strategy="epoch",
    save_strategy = "epoch",
    num_train_epochs = num_train_epochs,
    save_total_limit = num_train_epochs,
    per_device_train_batch_size = per_device_train_batch_size,
    gradient_accumulation_steps = gradient_accumulation_steps,
    # optim = optim,
    # save_steps=save_steps,
    logging_steps = logging_steps,
    learning_rate = learning_rate,
    weight_decay = weight_decay,
    fp16 = fp16,
    bf16 = bf16,
    max_grad_norm = max_grad_norm,
    max_steps = max_steps,
    warmup_ratio = warmup_ratio,
    group_by_length = group_by_length,
    lr_scheduler_type = lr_scheduler_type,
    report_to = "tensorboard",
)

#early stop
early_stopping = EarlyStoppingCallback(
    early_stopping_patience=3, # val损失3个epoch没有提升自动停止
    early_stopping_threshold=0.0
)

# Set supervised fine-tuning parameters
trainer = SFTTrainer(
    model = model,
    train_dataset = dataset['train'],
    eval_dataset = dataset["test"],
    # dataset_text_field = "text",
    # max_seq_length = max_seq_length,
    # tokenizer = tokenizer,
    args = training_arguments,
    # packing = packing,
    callbacks=[early_stopping],
    compute_metrics=compute_metrics,
)

class SaveBestModelCallback(TrainerCallback):
    def __init__(self):
        super().__init__()
        self.best_eval_loss = float('inf')
        self.best_model_checkpoint = None
        
    def on_log(self, args, state, control, logs=None, **kwargs):
        # Check if training_loss is in the logs and print it
        if 'loss' in logs:
            training_loss = logs['loss']
            logging.info(f"Epoch: {int(state.epoch)}, Step: {state.global_step}, Current training_loss: {training_loss}")

    # def on_evaluate(self, args, state, control, **kwargs):
    #     # Check if eval_loss is in the logs
    #     if 'eval_loss' in state.log_history[-1]:
    #         eval_loss = state.log_history[-1]['eval_loss']
    #         # Print current eval_loss with epoch and step
    #         logging.info(f"Epoch: {int(state.epoch)}, Step: {state.global_step}, Current eval_loss: {eval_loss}")  

    #         if eval_loss < self.best_eval_loss:
    #             self.best_eval_loss = eval_loss
    #             # Save the best model
    #             self.best_model_checkpoint = state.global_step
    #             trainer.save_model(f"{args.output_dir}/best_model")
    #             # Print loss of Best Model
    #             logging.info(f"New best model saved at step {state.global_step} with eval_loss: {eval_loss}")  

# Create an instance of the callback
save_best_model_callback = SaveBestModelCallback()

# Training and logging
logging.basicConfig(filename=output_dir+'/training.log', level=logging.INFO)
logging.info(f"""[Device]: cuda:{os.environ["CUDA_VISIBLE_DEVICES"]}...\n""")
logging.info(f"""[Model]: Loading {model_name}...\n""")
logging.info(f"""[Outputdir]: Loading {output_dir}...\n""")

# Add the callback to the trainer
trainer.add_callback(save_best_model_callback)

# Train model
trainer.train()

# Save trained model
trainer.model.save_pretrained(new_model_dir, max_shard_size="10GB")
